app/domain/authentication/authn_azure/verify_and_decode_azure_token.rb

```diff
@@ -0,0 +1,88 @@
+require 'jwt'
+
+module Authentication
+  module AuthnAzure
+
+    Err = Errors::Authentication::AuthnAzure
+    Log = LogMessages::Authentication::AuthnAzure
+
+    # We use the OIDC log messages for Identity Provider related code
+    OIDCLog = LogMessages::Authentication::AuthnAzure
+
+    VerifyAndDecodeAzureToken = CommandClass.new(
+      dependencies: {
+        # We have a ConcurrencyLimitedCache which wraps a RateLimitedCache which wraps a FetchProviderCertificate class
+        fetch_provider_certificate: ::Util::ConcurrencyLimitedCache.new(
+          ::Util::RateLimitedCache.new(
+            ::Authentication::AuthnOidc::FetchProviderCertificate.new,
+            refreshes_per_interval: 10,
+            rate_limit_interval:    300, # 300 seconds (every 5 mins)
+            logger: Rails.logger
+          ),
+          max_concurrent_requests: 3,
+          logger: Rails.logger
+        ),
+        logger:                     Rails.logger
+      },
+      inputs:       %i(provider_uri token_jwt)
+    ) do
+
+      def call
+        fetch_certs
+        verify_and_decode_azure_token
+      end
+
+      private
+
+      def fetch_certs(force_read: false)
+        @certs = @fetch_provider_certificate.(
+          provider_uri: @provider_uri,
+          refresh: force_read
+        )
+        @logger.debug(OIDCLog::IdentityProviderCertificateFetchedFromCache.new)
+        @certs
+      end
+
+      def verify_and_decode_azure_token
+        # We ensure that the certs are fresh just before we decode the token
+        ensure_certs_are_fresh
+
+        # `decoded_token` verifies the token before decoding it
+        decoded_token
+      end
+
+      def ensure_certs_are_fresh
+        decoded_token
+      rescue
+        @logger.debug(OIDCLog::ValidateIdentityProviderCertificateIsUpdated.new)
+        # maybe failed due to certificate rotation. Force cache to read it again
+        fetch_certs(force_read: true)
+      end
+
+      # The `JWT.decode` method verifies the token before decoding it
+      def decoded_token
+        return @decoded_token unless @decoded_token.nil?
+
+        # TODO: decide which claims we should verify
+        # specify the claims we need to verify
+        options = { verify_iss: true,
+                    iss: @provider_uri,
+                    verify_aud: true
+        }
+
+        @decoded_token = JWT.decode(
```

Comment:
as we use JWT in multiple places it might be good to have a dedicated object to deal with JWT related actions. then we can have one place to have defaults and change implementation if we would want in the future

Reply:
test

Reply:
Where exactly do we have `JWT` other than this place?